pkg/icingadb/history/sync.go

```diff
@@ -32,274 +37,359 @@ func NewSync(db *icingadb.DB, redis *icingaredis.Client, logger *zap.SugaredLogg
-	structifiers []structify.MapStructifier
-}
+// userNotificationStage is a specialized stageFunc that populates the user_notification_history table. It is executed
+// on the notification history stream and uses the users_notified_ids attribute to create an entry in the
+// user_notification_history relation table for each user ID.
+func userNotificationStage(ctx context.Context, s Sync, key string, in <-chan redis.XMessage, out chan<- redis.XMessage) error {
+	type NotificationHistory struct {
+		Id            types.UUID   `structify:"id"`
+		EnvironmentId types.Binary `structify:"environment_id"`
+		EndpointId    types.Binary `structify:"endpoint_id"`
+		UserIds       types.String `structify:"users_notified_ids"`
+	}
 
-// historyStreams contains all Redis history streams to sync.
-var historyStreams = func() []historyStream {
-	var streams []historyStream
-	for _, rhs := range []struct {
-		kind       string
-		structPtrs []v1.UpserterEntity
-	}{
-		{"notification", []v1.UpserterEntity{(*v1.NotificationHistory)(nil), (*v1.HistoryNotification)(nil)}},
-		{"usernotification", []v1.UpserterEntity{(*v1.UserNotificationHistory)(nil)}},
-		{"state", []v1.UpserterEntity{(*v1.StateHistory)(nil), (*v1.HistoryState)(nil)}},
-		{"downtime", []v1.UpserterEntity{(*v1.DowntimeHistory)(nil), (*v1.HistoryDowntime)(nil)}},
-		{"comment", []v1.UpserterEntity{(*v1.CommentHistory)(nil), (*v1.HistoryComment)(nil)}},
-		{"flapping", []v1.UpserterEntity{(*v1.FlappingHistory)(nil), (*v1.HistoryFlapping)(nil)}},
-		{"acknowledgement", []v1.UpserterEntity{(*v1.AcknowledgementHistory)(nil), (*v1.HistoryAck)(nil)}},
-	} {
-		var structifiers []structify.MapStructifier
-		for _, structPtr := range rhs.structPtrs {
-			structifiers = append(structifiers, structify.MakeMapStructifier(reflect.TypeOf(structPtr).Elem(), "json"))
+	structifier := structify.MakeMapStructifier(reflect.TypeOf((*NotificationHistory)(nil)).Elem(), "structify")
+
+	return writeMultiEntityStage(func(entry redis.XMessage) ([]v1.UpserterEntity, error) {
+		rawNotificationHistory, err := structifier(entry.Values)
+		if err != nil {
+			return nil, err
 		}
+		notificationHistory := rawNotificationHistory.(*NotificationHistory)
 
-		streams = append(streams, historyStream{rhs.kind, structifiers})
-	}
+		if !notificationHistory.UserIds.Valid {
+			return nil, nil
+		}
 
-	return streams
-}()
+		var users []types.Binary
+		err = internal.UnmarshalJSON([]byte(notificationHistory.UserIds.String), &users)
+		if err != nil {
+			return nil, err
+		}
+
+		var userNotifications []v1.UpserterEntity
+
+		for _, user := range users {
+			userNotifications = append(userNotifications, &v1.UserNotificationHistory{
+				EntityWithoutChecksum: v1types.EntityWithoutChecksum{
+					IdMeta: v1types.IdMeta{
+						Id: utils.Checksum(append(append(append([]byte(nil),
+							notificationHistory.EnvironmentId...), notificationHistory.Id.UUID[:]...), user...)),
```

Comment:
For the sake of completeness: Please remove the environment ID from the checksum and also from the schema comments.